Add  friction coefficient to InteractiveViewer

frankteller-de opened the pull request:
It would be nice if the kDrag parameter were public to change the scrolling behavior. 

For example I've changed the kDrag var into a public parameter called scrollFriction.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

frankteller-de committed:
scrollFriction added to InteractiveViewer

frankteller-de committed:
Update interactive_viewer.dart

spaces removed

alestiago commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -296,6 +300,16 @@ class InteractiveViewer extends StatefulWidget {
   /// than maxScale.
   final double minScale;
 
+  /// Changes the scrolling behavior after end of a gesture.
+  ///
+  /// Used as the coefficient of friction in the inertial translation animation.
+  /// This value was eyeballed to give a feel similar to Google Photos.
+  ///
+  /// Defaults to 0.0000135.
+  ///
+  /// Cannot be null, and must be a finite number greater than zero
```
```suggestion
  /// Cannot be null, and must be a finite number greater than zero.
```

alestiago replied:
Can we add documentation on how the scroll changes as the value increases or decreases?

alestiago commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -296,6 +300,16 @@ class InteractiveViewer extends StatefulWidget {
+  /// Defaults to 0.0000135.
+  ///
+  /// Cannot be null, and must be a finite number greater than zero
+  final double scrollFriction;
```
What about renaming to:

```suggestion
  final double scrollFrictionCoefficient;
```
The original documentation states "coefficient of friction". I believe physically this is different from just "friction". The coefficient of friction is usually represented with a value between 0 and 1.5 (https://www.engineeringtoolbox.com/friction-coefficients-d_778.html). Whereas friction is the force usually measured in Newtons.

Wdyt of this rename?
---
Resources:

From https://www.britannica.com/science/coefficient-of-friction:
> coefficient of friction, ratio of the frictional force resisting the motion of two surfaces in contact to the normal [force](https://www.britannica.com/science/force-physics) pressing the two surfaces together. It is usually symbolized by the Greek letter mu (μ). Mathematically, μ = F/N, where F is the frictional force and N is the normal force. B

It defers from just "friction" (Definitions from [Oxford Languages](https://languages.oup.com/google-dictionary-en):
> the resistance that one surface or object encounters when moving over another.

Here a comparison https://www.khanacademy.org/science/high-school-physics/two-dimensional-motion-2/friction/a/friction-ap-physics-1.

alestiago commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -296,6 +300,16 @@ class InteractiveViewer extends StatefulWidget {
+  /// Used as the coefficient of friction in the inertial translation animation.
+  /// This value was eyeballed to give a feel similar to Google Photos.
+  ///
+  /// Defaults to 0.0000135.
```
What if we define a const variable locally in this file, named something along the lines "_kDefaultScrollFrictionCoefficient". That way, if the value is updated the documentation can not be out of sync.

```dart
/// This value was eyeballed to give a feel similar to Google Photos.
_kDefaultScrollFrictionCoefficient = 0.0000135;
```

```suggestion
  /// Defaults to [_kDefaultScrollFrictionCoefficient].
```

alestiago commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -132,6 +135,7 @@ class InteractiveViewer extends StatefulWidget {
        assert(builder != null),
        assert(minScale != null),
        assert(minScale > 0),
+       assert(scrollFriction > 0),
```
```suggestion
       assert(scrollFrictionCoefficient > 0),
```

Adding this suggestion in case you agree with https://github.com/flutter/flutter/pull/109443/files#r944361349, in this case you can use GitHub to batch the suggestions and automatically make the change.

alestiago commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -120,6 +122,7 @@ class InteractiveViewer extends StatefulWidget {
     // use cases.
     this.maxScale = 2.5,
     this.minScale = 0.8,
+    this.scrollFriction = 0.0000135,
```
```suggestion
    this.scrollFrictionCoefficient = 0.0000135,
```

Adding this suggestion in case you agree with https://github.com/flutter/flutter/pull/109443/files#r944361349, in this case you can use GitHub to batch the suggestions and automatically make the change.

alestiago commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -74,6 +74,7 @@ class InteractiveViewer extends StatefulWidget {
     // use cases.
     this.maxScale = 2.5,
     this.minScale = 0.8,
+    this.scrollFriction = 0.0000135,
```
```suggestion
    this.scrollFrictionCoefficient= 0.0000135,
```
Adding this suggestion in case you agree with https://github.com/flutter/flutter/pull/109443/files#r944361349, in this case you can use GitHub to batch the suggestions and automatically make the change.

alestiago commented:
I agree with this change, it would be nice if the constant internal value could be changed. I think the solution is quite simple. Thanks for raising the PR!

I just added small nit comments regarding naming and documentation, overall looks good to me. Are you planning to add tests?

codeforce-dev commented:
Thank you @alestiago! I can add some tests but first I have to look at the documentation what I have to do to add a test. It was my first PR.

frankteller-de committed:
Apply suggestions from code review

scrollFriction renamed to scrollFrictionCoefficient

Co-authored-by: Alejandro Santiago <dev@alestiago.com>

frankteller-de committed:
Update interactive_viewer.dart

rename scrollFriction

frankteller-de committed:
update

frankteller-de committed:
update

frankteller-de committed:
* added _kDefaultScrollFrictionCoefficient
* added a test

frankteller-de committed:
Merge branch 'flutter:master' into master

codeforce-dev commented:
I've added a test but I don't know what I should test exactly. That parameter is very random.

frankteller-de committed:
Update interactive_viewer_test.dart

spaces removed

alestiago commented:
You could test that the amount of scroll differs when a value x is provided compared to when a value y is given.

justinmc reviewed (commented):
Thanks for opening this PR, I'm on board but I want to make sure there's not a more generic way to solve this.

This makes me think of Scrollable's [physics](https://master-api.flutter.dev/flutter/widgets/Scrollable/physics.html) parameter.  What if we allowed users to pass in the entire Simulation and defaulted to the current FrictionSimulation?  Or even more high level, maybe InteractiveViewer should be using ScrollPhysics?

@Piinks You're working on 2D scrolling, are you using regular ScrollPhysics for that, or is there anything that could also apply here?

I'm not saying that scrollFrictionCoefficient isn't the best solution here, I just want to think through these other options.

alestiago commented:
I agree with @justinmc making use of a subclass of ScrollPhysics to achieve the effect is an interesting and more consistent solution with the framework and other Scrollables.

I'm curious about @Piinks insights and opinion on this PR to 👀

frankteller-de committed:
scrollFrictionCoefficient test update

frankteller-de committed:
Merge branch 'master' of https://github.com/codeforce-dev/flutter-v3.0.5

Piinks commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -74,6 +77,7 @@ class InteractiveViewer extends StatefulWidget {
     // use cases.
     this.maxScale = 2.5,
     this.minScale = 0.8,
+    this.scrollFrictionCoefficient = _kDefaultScrollFrictionCoefficient,
```
This feels a bit misleading, InteractiveViewer is not a scrolling widget. If we use scrolling terms, users expect other scrolling conventions to apply, like having a ScrollPosition, and ScrollNotifications.

codeforce-dev replied:
It's a good point. How about we call it interactionEndFrictionCoefficient or dragEndFrictionCoefficient?

Piinks replied:
That sounds good to me! :)

Piinks reviewed (commented):
I would love to incorporate more scrolling into interactive viewer, but I think it is more complicated than adding some properties that emulate scrolling. I've been exploring this in https://github.com/flutter/flutter/issues/64210. Ultimately, under the hood InteractiveViewer is not a scrolling widget. We may be able to rework it to become one, but that may require breaking changes, which is why I have been working on a separate 2D scrollable that would avoid breaking existing code.

I don't think adding this is a bad idea, I would just avoid referring to it as scrolling.

frankteller-de committed:
scrollFrictionCoefficient changed to interactionEndFrictionCoefficient

frankteller-de committed:
scrollFrictionCoefficient changed to interactionEndFrictionCoefficient

frankteller-de committed:
update

frankteller-de committed:
update

justinmc commented:
Thanks @Piinks.

So using ScrollPhysics is off the table, but does it make sense to allow customizing the entire Simulation ([here](https://github.com/flutter/flutter/pull/109443/files#diff-53cba322695cdee672c77a688c05f785d4a1ddc91a2d1904b9a666b01dd4b991R881))?  Or maybe that would still be far too complicated and we should just stick with the coefficient?

Piinks commented:
> does it make sense to allow customizing the entire Simulation ([here](https://github.com/flutter/flutter/pull/109443/files#diff-53cba322695cdee672c77a688c05f785d4a1ddc91a2d1904b9a666b01dd4b991R881))? Or maybe that would still be far too complicated and we should just stick with the coefficient?

@codeforce-dev is there an open issue that this is seeking to resolve? It would be very helpful in answering these questions and better understanding the request.

codeforce-dev commented:
The coefficient would be the simplest way to optimize the "scrolling" behaviour for some purposes. A more complex solution is not necessary. For example, [pdf_render](https://pub.dev/packages/pdf_render) uses InterativeViewer to handle the user interaction. But in my opinion, the default drag behaviour with a value of 0.0000135 feels not good for that purpose. 

FrictionCoefficient
Screen 1 = 0.0000135 | Screen 2 = 0.01
<img src="https://user-images.githubusercontent.com/3636208/185345990-824696e6-43c4-4812-b2cf-0bfb4db8415a.gif" width="300" align="left"><img src="https://user-images.githubusercontent.com/3636208/185345997-7f9b41df-51dc-43d5-9a1e-1fea36d700b8.gif" width="300" align="left">

justinmc reviewed (commented):
It sounds to me like a custom Simulation is too confusing and low-level for use cases like the one [given by](https://github.com/flutter/flutter/pull/109443#issuecomment-1219206562)  @codeforce-dev.  I say let's just go forward with exposing the coefficent of friction like in this PR 👍.

A few comments below but I think this is almost ready.  Thanks for dealing with all of my questions on making sure this is the best solution!

frankteller-de committed:
Update interactive_viewer.dart

_kDefaultInteractionEndFrictionCoefficient constant removed for more readability

codeforce-dev commented on packages/flutter/test/widgets/interactive_viewer_test.dart:
```diff
@@ -1377,6 +1377,42 @@ void main() {
+      await tester.dragFrom(const Offset(100.0, 100.0), const Offset(0.0, 50.0));
+      final Vector3 translation2 = transformationController.value.getTranslation();
+
+      expect(translation2.y, equals(0));
```
Thanks for your reply. I've already tried dragFrom and moveBy with pumpAndSettle but unfortunately it has no effect on it. I always get -50.

```
final TestGesture gesture = await tester.startGesture(const Offset(100, 100));
await gesture.moveBy(const Offset(0, -50));
await tester.pumpAndSettle();
final Vector3 translation = transformationController.value.getTranslation();
// translation.y = -50;
```

codeforce-dev replied:
Sorry, the solution is moveTo instead moveBy.

codeforce-dev replied:
I think flingFrom is the solution.

frankteller-de committed:
interactionEndFrictionCoefficient test modified

frankteller-de committed:
Merge branch 'flutter:master' into master

Piinks commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -296,6 +300,16 @@ class InteractiveViewer extends StatefulWidget {
   /// than maxScale.
   final double minScale;
 
+  /// Changes the scrolling behavior after end of a gesture.
```
Nit: 

```suggestion
  /// Changes the deceleration behavior after a gesture.
```

Just avoiding future confusion about the interactive viewer not actually being a scrolling widget :)

Piinks commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -296,6 +300,16 @@ class InteractiveViewer extends StatefulWidget {
+  /// Changes the scrolling behavior after end of a gesture.
+  ///
+  /// Used as the coefficient of friction in the inertial translation animation.
+  /// This value was eyeballed to give a feel similar to Google Photos.
```
Can you remove this? This could change at any time, but also does not account for how Google Photos behaves on different devices.

justinmc replied:
+1.  Keep it in the comments below, but remove it from the public docs.  You could make it "Google Photos on Android 12" since I think that's what I was using when I wrote that comment.

Piinks reviewed (approved):
LGTM with a few nits - thank you!

justinmc commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -519,10 +533,6 @@ class _InteractiveViewerState extends State<InteractiveViewer> with TickerProvid
 
-  // Used as the coefficient of friction in the inertial translation animation.
-  // This value was eyeballed to give a feel similar to Google Photos.
-  static const double _kDrag = 0.0000135;
```
Can you actually keep this constant and the comments here?  And then use the constant as the default value above instead of hard coding it.

justinmc commented on packages/flutter/test/widgets/interactive_viewer_test.dart:
```diff
@@ -1377,6 +1377,59 @@ void main() {
+        MaterialApp(
+          home: Scaffold(
+            body: SizedBox(
+              width: 200, height: 200,
```
Separate these onto different lines.

justinmc commented on packages/flutter/test/widgets/interactive_viewer_test.dart:
```diff
@@ -1377,6 +1377,59 @@ void main() {
+      await tester.pumpAndSettle();
+      final Vector3 translation2 = transformationController2.value.getTranslation();
+
+      // Test 2 movement must be greater than in Test 1.
```
Change this to: The coefficient 0.01 is greater than the default of 0.0000135, so the translation comes to a stop more quickly.

justinmc reviewed (approved):
LGTM with nits as well.  Let me know if you want to fix those and we can get this merged. :+1:

Piinks commented:
Hi @codeforce-dev, would you like to address the outstanding comments from @justinmc so this can land? Thanks!

frankteller-de committed:
update

frankteller-de committed:
Merge branch 'flutter:master' into master

frankteller-de committed:
Merge branch 'master' of https://github.com/codeforce-dev/flutter-v3.0.5

frankteller-de committed:
Merge branch 'flutter:master' into master

Piinks committed:
Update packages/flutter/lib/src/widgets/interactive_viewer.dart
